### ISSUE TITLE
KOL/KUAL: freeze of touch right after Kindle reboot

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#11199]
* KOReader version: v2023.10-37, -41, -55
* Device: Kindle Scribe (5.16.2)

#### Issue and Steps to reproduce
I noticed that frequently (about 6/10 times) this happens:
1) I reboot Kindle normally
2) Wait until home screen shows up
3) Very soon (within 3-5 seconds once it is loaded) hit KOL 1.3.6 icon to start KOReader in normal mode directly.
4) KOReader shows up OK, opens the book.
5) However, touch input is not working, both taps and swipes. I can press power button - usual screensaver shows up, press power button again - it wakes up, so KOReader and the system are not frozen, just the input lost somewhere in between those two. SSH works as well, but `lipc` command to call home doesn't work, probably `awesome` & Co. are shutdown at the moment.

I then just end up hard restarting by power button. Upon boot, I need to re-enable device passcode, because KOReader was shutdown and didn't have a chance to restore it.

I tried with and without USBnet - by uninstalling it via mrpi (which every time worked, but then a new KPV-something-something error log showed up in the /documents folder, but that's another story :D).

Launching from KUAL - KOReader menu works OK every time, so I am pretty sure it is KOL.

##### `crash.log` (if applicable)
KOReader's crash.log (this one is from yesterday, but today's "freezes" were about the same: 
[crash_KOL_startup_freeze.txt](https://github.com/koreader/koreader/files/13541244/crash_KOL_startup_freeze.txt)

In these cases, it always ends at this line:
```
12/02/23-23:28:35 INFO  Restoring user input handling 
```

FWIW, here is the `/var/log/messages` file that I backed up: 
[messages.txt](https://github.com/koreader/koreader/files/13541261/messages.txt)

**P.S.**
After all this and when I found that KUAL-launch works OK every time, I removed KOL for now. Then also disabled Amazon logging. I plan to just read upcoming weekdays, but I can come back to logs next weekend and try to reproduce if needed.

Please let me know what else I can provide in the meanwhile.

**Comment by NiLuJe** [comment]
Do you start with the last book open?

AFAIK, KOL isn't packaged for any of the new devices, and I don't recall if it's doing anything device-specific (e.g., the Voyage haptic buttons trickery), so... here be dragons? ^^.

**Comment by mergen3107** [comment]
I don't know if that's related and don't remember relative time, but recently I was trying to use OpenSSH instead of dripbear in usbnet. 

Kindle didn't seem to be too happy about this on reboot, saying "Starting sshd" as usual first, then "N E failed to stop sshd", but then starting it OK in three rows of text anyway. 

I then disabled it.

I _think_ the KOL issue started before OpenSSH, but definitely after I started using Enable ssh on boot, since I turned that on almost immediately :D 
But then again it is reproducible even with usbnet uninstalled via mrpi.

**Comment by NiLuJe** [comment]
> Kindle didn't seem to be too happy about this on reboot, saying "Starting sshd" as usual first, then "N E failed to stop sshd", but then starting it OK in three rows of text anyway.

IIRC, that's the expected behavior after any kind of reboot where SSH was not disabled manually (which is pretty much all of them ;p).

**Comment by mergen3107** [comment]
People from 4PDA forum confirmed the issue. 

1) 5.14.X with previous KOL 1.3 is OK. 
2) Kindle Scribe 5.16.2 with KOL 1.3.6 (regular mode) looses touch input only when opening KOReader with KOL for the first time right after reboot. If you open it in KUAL first, or wait quite some time before launching via KOL, no issues with subsequent launches during the same boot session.

**Comment by yparitcher** [comment]
1. Does this happen with the frameworkless branch?
2. Does it happen when you launch koreader  with KUAL `asap`? (what KOL uses)
3. You can check `/var/tmp/KOL.log`

Otherwise my guess is a timing / race issue as KOL runs & exits much quicker, not giving the native system time to settle. You can try putting a sleep in the beginning of `koreader.sh` to test this theory.

**Comment by mergen3107** [comment]
@yparitcher thanks! I'll check

**Comment by Frenzie** [comment]
A freeze doesn't sound like a nitpick? :)

**Comment by mergen3107** [comment]
Yep, because KUAL launch is OK. I also don't think I'll have much time trying to reproduce it again due to work schedule before New Year

**Comment by mergen3107** [comment]
Welp, I just got a freeze by launching from KUAL as well.
1) Turned off Wi-Fi
2) Restarted Kindle
3) Upon reboot, enter my pin code to enter Kindle
4) Went into KUAL immediately
5) Started KOReader - no touch!
6) Put to screensaver and back, touch didn't come back. Restarted again to get the crash.log.

crash.log: [crash_KUAL_freeze.txt](https://github.com/koreader/koreader/files/13776250/crash_KUAL_freeze.txt)

**Comment by mergen3107** [comment]
I can only see these repetetive errors:
```
error: Unable to read ICC workflow
warning: Attempt to read Output Intent failed
```
However, these should be related to `libmupdf.so` (did a quick search).

**Comment by NiLuJe** [comment]
Umm, I'm actually seeing input events going through starting at `12/27/23-01:41:39`?

**Comment by mergen3107** [comment]
Yes, that was me tapping on screen. It started with me doing top edge left gesture because I wanted to go back to the other book, but couldn't.

**Comment by NiLuJe** [comment]
My point is that if those get through, then we *have* working input, so, I'm... confused?

**Comment by mergen3107** [comment]
Ah, I see. These look like taps though, no swipes.
Let me see if I can reproduce again.
Maybe killing `pillow` and `awesome` very early doesn't let them do their usual stuff they are supposed to do after startup?

**Comment by NiLuJe** [comment]
I wouldn't expect those two to have anything to do with input, one is an X11 WM and the other a small GUI thingy ;).

**Comment by mergen3107** [comment]
Then it must be some of these :D 
```
# List of services we stop in order to reclaim a tiny sliver of RAM...
TOGGLED_SERVICES="stored webreader kfxreader kfxview todo tmd rcm archive scanner otav3 otaupd"
```

**Comment by NiLuJe** [comment]
Unlikely either, I suspect it's just a weird timing quirk; possibly a resurgence of the weird Voyage feature that allowed the framework to neuter the haptic buttons...

**Comment by mergen3107** [comment]
I caught it again!
Turns out, this can mostly (only?) happen in Airplane mode.
With Wi-Fi on, I did 5 restarts with no issue.
Let's what's going on with usbnet...

**Comment by mergen3107** [comment]
Here is just after it started stuck: [crash_KUAL_freeze_2_short.txt](https://github.com/koreader/koreader/files/13780337/crash_KUAL_freeze_2_short.txt)

This is me trying to touch with fingers, then with pen: [crash_KUAL_freeze_2_long_with_touch_and_pen.txt](https://github.com/koreader/koreader/files/13780343/crash_KUAL_freeze_2_long_with_touch_and_pen.txt)

**Comment by mergen3107** [comment]
Kindle responds to gyro rotations alright, even the clock in the footer updates on time :D

**Comment by mergen3107** [comment]
> allowed the framework to neuter the haptic buttons...

Do you have any referenced issue describing this? I'll go look Voyage device:init, maybe there were some workarounds.

**Comment by mergen3107** [comment]
> I caught it again!
> Turns out, this can mostly (only?) happen in Airplane mode.
> With Wi-Fi on, I did 5 restarts with no issue.
> Let's what's going on with usbnet...

P.S. It seems that KOL itself is harmless, it just allowed to start KOReader way earlier than KUAL.

**Comment by mergen3107** [comment]
1) Go to home screen with `lipc-set-prop com.lab126.appmgrd start app://com.lab126.booklet.home`:
nothing happened

2) `killall reader.lua`:
Home screen shows up, no top bar, touch works.

3) Tap KOL:
goes into KOReader without touch again.

**Comment by mergen3107** [comment]
> 2\. no top bar

This one is fixed by `lipc-set-prop com.lab126.pillow disableEnablePillow enable`. It comes back and it works.

**Comment by mergen3107** [comment]
> 1. Go to home screen with `lipc-set-prop com.lab126.appmgrd start app://com.lab126.booklet.home`:
>    nothing happened

OK, so doing the whole thing from `koreader.sh` works OK:
```
if [ "${AWESOME_STOPPED}" = "yes" ]; then
        logmsg "Resuming awesome . . ."
        killall -CONT awesome
    fi
    if [ "${PILLOW_HARD_DISABLED}" = "yes" ]; then
        logmsg "Enabling pillow . . ."
        # NOTE: Try to leave the user with a slightly more useful FB content than our own last screen...
        cat /var/tmp/koreader-fb.dump >/dev/fb0
        rm -f /var/tmp/koreader-fb.dump
        lipc-set-prop com.lab126.pillow disableEnablePillow enable
        lipc-set-prop com.lab126.appmgrd start app://com.lab126.booklet.home
    fi
```

**Comment by mergen3107** [comment]
So, it seems touch itself works in Kindle, but not when KOReader is started or restarted in _that_ particular framework session.
Before I do `framework stop` and `framework start`, is there anything I should check?

It might take me to a while to catch this again, so I wanted to squeeze out most of things while I am here.

**Comment by mergen3107** [comment]
> double-check that all the input devices got enumerated in the expected places, and that they actually spit out useful data over SSH via evtest or something.

While in KOReader, I did `evtest /dev/input/by-path/platform-1001e000.i2c-event` and it showed this (no updates upon neither finger or pen input):
```
  Event type 0 (Sync)
  Event type 1 (Key)
    Event code 320 (ToolPen)
    Event code 321 (ToolRubber)
    Event code 330 (Touch)
    Event code 331 (Stylus)
    Event code 332 (Stylus2)
  Event type 3 (Absolute)
    Event code 0 (X)
      Value   1613
      Min        0
      Max    15624
    Event code 1 (Y)
      Value   8439
      Min        0
      Max    20832
    Event code 24 (Pressure)
      Value      0
      Min        0
      Max     4095
    Event code 25 (Distance)
      Value     62
      Min        0
      Max      255
    Event code 26 (XTilt)
      Value   -700
      Min    -9000
      Max     9000
    Event code 27 (YTilt)
      Value  -2600
      Min    -9000
      Max     9000
Testing ... (interrupt to exit)
```
Outside KOReader, this command only responds to the pen like this (it spams a LOT of messages):
```
...
Event: time 1703703206.777027, type 3 (Absolute), code 0 (X), value 7951
Event: time 1703703206.777027, type 3 (Absolute), code 1 (Y), value 20680
Event: time 1703703206.777027, type 3 (Absolute), code 27 (YTilt), value -1700
Event: time 1703703206.777027, -------------- Report Sync ------------
Event: time 1703703206.779786, type 3 (Absolute), code 0 (X), value 7959
Event: time 1703703206.779786, type 3 (Absolute), code 1 (Y), value 20688
Event: time 1703703206.779786, type 3 (Absolute), code 25 (Distance), value 23
Event: time 1703703206.779786, -------------- Report Sync ------------
...
```

**Comment by mergen3107** [comment]
Doing `evtest /dev/input/touch` (which is a link to `/dev/input/event2`), shows finger touches on updates in both KOReader and Kindle home:
```
[root@kindle root]# evtest /dev/input/touch
Input driver version is 1.0.1
Input device ID: bus 0x0 vendor 0x0 product 0x0 version 0x0
Input device name: "pt_mt"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
  Event type 2 (Relative)
  Event type 3 (Absolute)
    Event code 47 (MT Slot)
      Value      0
      Min        0
      Max        1
    Event code 48 (MT Major Axis Ellipse)
      Value      0
      Min        0
      Max      255
    Event code 53 (MT X)
      Value      0
      Min        0
      Max     1859
    Event code 54 (MT Y)
      Value      0
      Min        0
      Max     2479
    Event code 55 (MT Tool type)
      Value      0
      Min        0
      Max        2
    Event code 57 (MT Tracking ID)
      Value      0
      Min        0
      Max    65535
    Event code 58 (Pressure on contact area)
      Value      0
      Min        0
      Max      255
    Event code 59 (Contact hover distance)
      Value      0
      Min        0
      Max      255
Testing ... (interrupt to exit)
Event: time 1703703436.220036, type 3 (Absolute), code 57 (MT Tracking ID), value 0
Event: time 1703703436.220036, type 3 (Absolute), code 53 (MT X), value 876
Event: time 1703703436.220036, type 3 (Absolute), code 54 (MT Y), value 1166
Event: time 1703703436.220036, type 3 (Absolute), code 58 (Pressure on contact area), value 106
Event: time 1703703436.220036, -------------- Report Sync ------------
Event: time 1703703436.243662, type 3 (Absolute), code 53 (MT X), value 875
Event: time 1703703436.243662, type 3 (Absolute), code 58 (Pressure on contact area), value 116
Event: time 1703703436.243662, -------------- Report Sync ------------
```

**Comment by NiLuJe** [comment]
> Do you have any referenced issue describing this? I'll go look Voyage device:init, maybe there were some workarounds.

Nope, it's a flag in the appreg db set during KUAL install on affected devices, so it's in the KUAL install script.

> then with pen

So pen & gyro work, lol.

**Comment by mergen3107** [comment]
OK, progress :D Weird, though, let's discuss.
This line refers to the device I did evtest first:
https://github.com/koreader/koreader/blob/c12b4f2e14c83b982fcecaa6ca5d1a178c9e4a1a/frontend/device/kindle/device.lua#L641

1) I changed it to `touch_dev = "/dev/input/touch",` and KOReader upon restart worked with touch!
2) Can I leave it as `touch_dev = "/dev/input/touch",` or should I make it `touch_dev = "/dev/input/event2",` (which is the `touch` link's target), or use something else?
3) Why on earth `/dev/input/by-path/platform-1001e000.i2c-event` worked at all? :D This must be just the Wacom digitizer. Unless, Kindle was doing something smart on startup, relaying the pen input as regular touch input.

**Comment by mergen3107** [comment]
> So pen & gyro work, lol.

In "frozen" status, only Gyro worked.

**Comment by NiLuJe** [comment]
> In "frozen" status, only Gyro worked.

I'm seeing *lots* of pen events in your log

**Comment by mergen3107** [comment]
> I'm seeing _lots_ of pen events in your log

Ah, yes, but I meant, they didn't actually do anything either.
This is in sync with what evtest showed while in KOReader.

**Comment by NiLuJe** [comment]
> 3\. Why on earth `/dev/input/by-path/platform-1001e000.i2c-event` worked at all? :D This must be just the Wacom digitizer. Unless, Kindle was doing something smart on startup, relaying the pen input as regular touch input.

Doesn't that point to `event2` too?

**Comment by mergen3107** [comment]
> Doesn't that point to `event2` too?

Nope, that points to `event3`

**Comment by NiLuJe** [comment]
What does evtest say the name of that one is?

(And do you get pen inputs via the `touch` one?)

**Comment by mergen3107** [comment]
Here is the whole tree with links:
```
[root@kindle by-path]# tree /dev/input/ -l
/dev/input/
├── by-path
│   ├── platform-10019000.i2c-platform-gpio-keys.7.auto-event -> ../event0
│   ├── platform-1001e000.i2c-event -> ../event3
│   └── platform-11007000.i2c-event-joystick -> ../event1
├── event0
├── event1
├── event2
├── event3
├── event4
├── stylus -> event3
├── touch -> event2
└── uinput
```

**Comment by NiLuJe** [comment]
Ohkay, so, definitely scrap the by-path stuff and add both touch & stylus instead, yeah ;).

**Comment by NiLuJe** [comment]
> > I'm seeing _lots_ of pen events in your log
> 
> Ah, yes, but I meant, they didn't actually do anything either - no swipes, no tap with pen did anything to KOReader UI. This is in sync with what evtest showed while in KOReader.

Ooh, looking closer, they weren't getting rescaled by the driver. That's.. interesting ;p.

**Comment by mergen3107** [comment]
> What does evtest say the name of that one is?

For the OG `/dev/input/by-path/platform-1001e000.i2c-event`, it said this (sorry, I didn't copy the first part in https://github.com/koreader/koreader/issues/11199#issuecomment-1870552183):
```
[root@kindle by-path]# evtest /dev/input/by-path/platform-1001e000.i2c-event
Input driver version is 1.0.1
Input device ID: bus 0x18 vendor 0x2d1f product 0x158 version 0x1827
Input device name: "WacomDigitizer"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
    Event code 320 (ToolPen)
    Event code 321 (ToolRubber)
    Event code 330 (Touch)
    Event code 331 (Stylus)
    Event code 332 (Stylus2)
  Event type 3 (Absolute)
    Event code 0 (X)
      Value  14072
      Min        0
      Max    15624
    Event code 1 (Y)
      Value  11149
      Min        0
      Max    20832
    Event code 24 (Pressure)
      Value      0
      Min        0
      Max     4095
    Event code 25 (Distance)
      Value     47
      Min        0
      Max      255
    Event code 26 (XTilt)
      Value  -2600
      Min    -9000
      Max     9000
    Event code 27 (YTilt)
      Value   -600
      Min    -9000
      Max     9000
Testing ... (interrupt to exit)
```

**Comment by NiLuJe** [comment]
> Ohkay, so, definitely scrap the by-path stuff and add both touch & stylus instead, yeah ;).

And hopefully the protocol doesn't get mangled further and everything still works as-is ;p.

**Comment by mergen3107** [comment]
In KOReader, `evtest /dev/input/touch` does not respond (update), but touch itself actually works.
Outside KOReader, it responds:
```
[root@kindle by-path]# evtest /dev/input/touch
Input driver version is 1.0.1
Input device ID: bus 0x0 vendor 0x0 product 0x0 version 0x0
Input device name: "pt_mt"
Supported events:
  Event type 0 (Sync)
  Event type 1 (Key)
  Event type 2 (Relative)
  Event type 3 (Absolute)
    Event code 47 (MT Slot)
      Value      0
      Min        0
      Max        1
    Event code 48 (MT Major Axis Ellipse)
      Value      0
      Min        0
      Max      255
    Event code 53 (MT X)
      Value      0
      Min        0
      Max     1859
    Event code 54 (MT Y)
      Value      0
      Min        0
      Max     2479
    Event code 55 (MT Tool type)
      Value      0
      Min        0
      Max        2
    Event code 57 (MT Tracking ID)
      Value      0
      Min        0
      Max    65535
    Event code 58 (Pressure on contact area)
      Value      0
      Min        0
      Max      255
    Event code 59 (Contact hover distance)
      Value      0
      Min        0
      Max      255
Testing ... (interrupt to exit)
Event: time 1703704418.481502, type 3 (Absolute), code 57 (MT Tracking ID), value 0
Event: time 1703704418.481502, type 3 (Absolute), code 53 (MT X), value 948
Event: time 1703704418.481502, type 3 (Absolute), code 54 (MT Y), value 2440
Event: time 1703704418.481502, type 3 (Absolute), code 58 (Pressure on contact area), value 84
Event: time 1703704418.481502, type 3 (Absolute), code 48 (MT Major Axis Ellipse), value 107
Event: time 1703704418.481502, -------------- Report Sync ------------
Event: time 1703704418.517362, type 3 (Absolute), code 54 (MT Y), value 2439
Event: time 1703704418.517362, type 3 (Absolute), code 58 (Pressure on contact area), value 101
Event: time 1703704418.517362, type 3 (Absolute), code 48 (MT Major Axis Ellipse), value 95
Event: time 1703704418.517362, -------------- Report Sync ------------
Event: time 1703704418.542654, type 3 (Absolute), code 54 (MT Y), value 2438
Event: time 1703704418.542654, type 3 (Absolute), code 58 (Pressure on contact area), value 100
Event: time 1703704418.542654, -------------- Report Sync ------------
Event: time 1703704418.567812, type 3 (Absolute), code 53 (MT X), value 949
Event: time 1703704418.567812, type 3 (Absolute), code 58 (Pressure on contact area), value 68
Event: time 1703704418.567812, type 3 (Absolute), code 48 (MT Major Axis Ellipse), value 71
Event: time 1703704418.567812, -------------- Report Sync ------------
Event: time 1703704418.592782, type 3 (Absolute), code 57 (MT Tracking ID), value -1
Event: time 1703704418.592782, -------------- Report Sync ------------

```

**Comment by NiLuJe** [comment]
We grab exclusive access to input devices on kindle, you won't be able to follow via evtest ;).

**Comment by mergen3107** [comment]
> Ohkay, so, definitely scrap the by-path stuff and add both touch & stylus instead, yeah ;).

So, should I combine them in here?
```
local KindleScribe = Kindle:extend{
    model = "KindleScribe",
    isMTK = yes,
    isTouchDevice = yes,
    hasFrontlight = yes,
    hasNaturalLight = yes,
    -- NOTE: We *can* technically control both LEDs independently,
    --       but the mix is device-specific, we don't have access to the LUT for the mix powerd is using,
    --       and the widget is designed for the Kobo Aura One anyway, so, hahaha, nope.
    hasNaturalLightMixer = yes,
    hasGSensor = yes,
    display_dpi = 300,
    -- touch_dev = "/dev/input/by-path/platform-1001e000.i2c-event",
    touch_dev = "/dev/input/touch",
    -- NOTE: TBC whether dithering actually works on Bellatrix3...
    canHWDither = no,
    canDoSwipeAnimation = yes,
}
```
or in the Kindle:init?

> We grab exclusive access to input devices on kindle, you won't be able to follow via evtest ;).

Good to know :D

**Comment by mergen3107** [comment]
> 3\. Why on earth `/dev/input/by-path/platform-1001e000.i2c-event` worked at all? :D This must be just the Wacom digitizer. Unless, Kindle was doing something smart on startup, relaying the pen input as regular touch input.

But this mistery persists lol

**Comment by NiLuJe** [comment]
> So, should I combine them in here?

Nope, the two are split, touch is in the first block, pen is in init ;).

**Comment by mergen3107** [comment]
> or in the Kindle:init?

In there, I already added the pen input though:
```
    self.input.open(self.touch_dev)
    self.input.open("fake_events")

    -- Setup pen input
    self.input.wacom_protocol = true
    self.input.open("/dev/input/event4")
```

**Comment by NiLuJe** [comment]
> But this mistery persists lol

That smells like the other bellatrix boards where that symlink also went poof, so I would'nt worry too much about it ;p

( #11049 & #11052)

**Comment by mergen3107** [comment]
> Nope, the two are split, touch is in the first block, pen is in init ;).

OK, so I can leave it as this? Or direct link to `event2`?
```
    touch_dev = "/dev/input/touch",
```

**Comment by mergen3107** [comment]
> bellatrix boards where that symlink also went poof

Ah, I see -_- lab126 just reused the hardware name for Wacom digitizer.

**Comment by mergen3107** [comment]
Question then, if the pen input was still defined correctly, why didn't it work?

**Comment by mergen3107** [comment]
The issue is reproducible, I reverted this now:
```
    touch_dev = "/dev/input/by-path/platform-1001e000.i2c-event",
    -- touch_dev = "/dev/input/touch",
```
and it doesn't work again. Good!

**Comment by mergen3107** [comment]
Here is full list and descriptions of tge input devices:
```
[root@kindle by-path]# cat /proc/bus/input/devices
I: Bus=0019 Vendor=0001 Product=0001 Version=0100
N: Name="bd71828-pwrkey"
P: Phys=gpio-keys/input0
S: Sysfs=/devices/platform/10019000.i2c/i2c-1/1-004b/gpio-keys.7.auto/input/input0
U: Uniq=
H: Handlers=event0 perfmgr
B: PROP=0
B: EV=3
B: KEY=100000 0 0 0

I: Bus=0018 Vendor=003d Product=0000 Version=0000
N: Name="kx132-accel"
P: Phys=
S: Sysfs=/devices/platform/11007000.i2c/i2c-0/0-001f/input/input1
U: Uniq=
H: Handlers=event1 perfmgr
B: PROP=0
B: EV=9
B: ABS=1000007

I: Bus=0000 Vendor=0000 Product=0000 Version=0000
N: Name="pt_mt"
P: Phys=2-0024/input0
S: Sysfs=/devices/platform/1001e000.i2c/i2c-2/2-0024/input/input2
U: Uniq=
H: Handlers=event2 perfmgr
B: PROP=2
B: EV=f
B: KEY=0
B: REL=0
B: ABS=ee18000 0

I: Bus=0018 Vendor=2d1f Product=0158 Version=1827
N: Name="WacomDigitizer"
P: Phys=
S: Sysfs=/devices/platform/1001e000.i2c/i2c-2/2-0009/input/input3
U: Uniq=
H: Handlers=event3 perfmgr
B: PROP=2
B: EV=b
B: KEY=1c03 0 0 0 0 0 0 0 0 0 0
B: ABS=f000003

I: Bus=0018 Vendor=2d1f Product=0158 Version=1827
N: Name="stylus-custom"
P: Phys=
S: Sysfs=/devices/virtual/input/input4
U: Uniq=
H: Handlers=event4 perfmgr
B: PROP=0
B: EV=b
B: KEY=1c03 0 0 0 0 0 0 0 0 0 0
B: ABS=f000003
```
I wonder if hardcoding touch input to `event2` is enough, or I should walk through `/sys/class/input` the same way as in #11052 ?

**Comment by NiLuJe** [comment]
> Question then, if the pen input was still defined correctly, why didn't it work?

The Wacom digitizer is usually much higher resolution than the panel, so there's a rescaling pass. Depending on the implementation, it can either be done in the driver (like it appears to be here when it works), or in userland (e.g., on rM).

I don't know how it all shakes up here, but I suspect everything works properly via the stylus symlink, probably because it's a virtual device or a shim that handles the extra processing ;).

**Comment by NiLuJe** [comment]
> I wonder if hardcoding touch input to `event2` is enough, or I should walk through `/sys/class/input` the same way as in #11052 ?

Nah, those touch & stylus symlinks sound very custom and purpose-built, so I doubt they'll change any time soon.

**Comment by mergen3107** [comment]
> but I suspect everything works properly via the stylus symlink

OK, so I will then replace this one as well:
```
    -- Setup pen input
    self.input.wacom_protocol = true
    self.input.open("/dev/input/event4")
```
to
```
    -- Setup pen input
    self.input.wacom_protocol = true
    self.input.open("/dev/input/stylus") -- sym link
```

**Comment by NiLuJe** [comment]
Yup! (unnecessary comment, though ;p)

**Comment by mergen3107** [comment]
@NiLuJe 
I am a bit confused by https://github.com/koreader/koreader/pull/11052/files
How does the code select which of these are correct touch input?

For example (https://github.com/koreader/koreader/issues/11199#issuecomment-1870577888), there are four non-0 `ABS`. How would it choose here? Or on PW5 there is only one non-0 `ABS`?

**Comment by mergen3107** [comment]
> unnecessary comment, though ;p

Yep) this was just to facilitate the comment (for passers by).

**Comment by NiLuJe** [comment]
Yeah, it would need an actual proper parsing of the cap bits on more complicated devices.

It works because there's nothing else to confuse it with, but, yeah, you'd probably pick up at least the gyro on something like a Scribe ;).

**Comment by mergen3107** [comment]
Interesting, `/dev/input/stylus` is weird: crash.log again see all its inputs, but they don't do anything (taps/swipes).

I should probably leave it there, but I just don't understand why `touch` symlink works, but `stylus` symlink doesn't.

**Comment by mergen3107** [comment]
So I didn't touch the pen input yet, because with `event4` it still works.

**Comment by mergen3107** [comment]
_Somehow_ I managed to take a screenshot (this is a tap at two opposite corners!) using the `stylus` input O_O. Maybe it didn't rescale again? But why?

**Comment by mergen3107** [comment]
> ├── stylus -> event3

OK, my bad!

`stylus` is in fact `event3`, not `event4`.

**Comment by mergen3107** [comment]
Thank you very much @NiLuJe!
This has been definitely one of the most interesting adventures with KOReader so far :D

**Comment by mergen3107** [comment]
Final test successful: five restarts+KOReader, no issues.
Waiting for PR to close this issue :)

**Comment by NiLuJe** [comment]
FWIW, I'd forgotten, but I just noticed that we explicitly already requested the virtual stylus ones instead of the raw digitizer (and, err, you're the one who tested it at the time ;p. c.f., #11159)

**Comment by mergen3107** [comment]
That was a _month_ ago, can't remember stuff these days :D

**Comment by mergen3107** [comment]
@yparitcher 
Sorry, a side question.
I found a file `/var/tmp/KOL.log`, and it has a bunch of:
```
stop: Unknown instance: 
sh: write error: Broken pipe
```

Is this normal or I can help you diagnose it?
Kindle Scribe, FW 5.16.2.0.1

**Comment by yparitcher** [comment]
> @yparitcher 
> Sorry, a side question.
> I found a file `/var/tmp/KOL.log`, and it has a bunch of:
> ```
> stop: Unknown instance: 
> sh: write error: Broken pipe
> ```
> 
> Is this normal or I can help you diagnose it?
> Kindle Scribe, FW 5.16.2.0.1

I see them in my log too,
The stderr of `koreader.sh` gets written to `KOL.log` so this is probably a case of one of the pipes in the shell script closing before the other end stops writing to it.
Which one i have no clue, but it appears to be harmless as koreader works normally.

TLDR
a koreader.sh (bug?), nothing to do with kol, you might see similar in kual.log

**Comment by mergen3107** [comment]
Thanks!
